llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td

```diff
@@ -2421,6 +2421,42 @@ let Predicates = [HasSVEorSME] in {
   defm FSQRT_ZPmZ  : sve_fp_2op_p_zd_HSD<0b01101, "fsqrt",  AArch64fsqrt_mt>;
 } // End HasSVEorSME
 
+// Helper for creating scalar fp -> int -> fp conversions using SVE.
```

Comment:
Rather than doing this in TableGen patterns, I was hoping that we could emulate the `FP_TO_INT` and `INT_TO_FP` with SVE instructions by using Custom lowering in AArch64ISelLowering, so that these operations are lowered using `LowerFixedLengthIntToFPToSVE` and `LowerFixedLengthFPToIntToSVE`. I would then expect a DAGCombine to fold away the scalar extracts/inserts, so that we'd end up with the same codegen as you get here, but without the need to explicitly pattern-match `FP_TO_INT(INT_TO_FP(..))`.

Reply:
When the streaming instructions are executed on a dedicated PE, then there is likely to be a significant cost for moving data between GPR and FPR/ZPR register files, keeping data in the same register file avoids that. Also, when a function is in streaming mode then there is the expectation that most operations will be SME or SVE instructions, so I'd expect this case (where the result is required in a GPR) to be rare. If this does happen then the cost of the extra `fmov` is probably not significant.

The specification for fcvtzs (scalar, integer) says:
> If FEAT_SME is implemented and the PE is in Streaming SVE mode, then any subsequent instruction which is dependent on the general-purpose register written by this instruction might be significantly delayed.

Reply:
So, I can reuse `LowerFixedLengthFPToIntToSVE()` and `LowerFixedLengthIntToFPToSVE()` by promoting the scalar conversions to legal fixed vector types in `LowerFP_TO_INT/INT_TO_FP()` + some inserts/extracts. The extracts/inserts do fold away, but you do get the side-effect that the predicate is for the full vector type (e.g. `ptrue p0.d, vl4`) for f32 (rather than `vl1`). I guess that should be fine though.

Reply:
Updated to lower `FP_TO_INT` and `INT_TO_FP` separately. I did not reuse `LowerFixedLengthFPToIntToSVE()` and `LowerFixedLengthIntToFPToSVE()`, but rather used the general SVE lowering. The ``LowerFixedLength...` methods are not really what we want here as they can introduce unnecessary packing/unpacking when the value we are about is only in lane 0.